### PR TITLE
feat(auth): Add TOTP support in Project and Tenant config

### PR DESCRIPTION
**Pull request by pragatimodi**
~**DO_NOT_MERGE** (waiting for BE changes)~

PR for TOTP Config for Project and Tenant

RELEASE NOTES: Added support for TOTP (Time-based One Time Password) as a Multi-Factor Authentication option in Tenant and Project config.

### CODE REVIEW

File: src/auth/auth-config.ts
Comment (prameshj) [inline src/auth/auth-config.ts:511]: Extend the comment with "new MFA providers (except phone) will indicate enablement/disablement through this field."

File: src/auth/auth-config.ts
Comment (prameshj) [inline src/auth/auth-config.ts:647]: nit: space between MultiFactorProviderConfig and Type.

File: src/auth/auth-config.ts
Comment (prameshj) [inline src/auth/auth-config.ts:650]: Should we instead check:

`if (typeof multiFactorProviderConfig.state === 'undefined' || ( multiFactorProviderConfig.state !== 'ENABLED' && multiFactorProviderConfig.state !== 'DISABLED'))`

File: src/auth/auth-config.ts
Comment (prameshj) [inline src/auth/auth-config.ts:658]: not specifying totpProviderConfig is an error right now, since we don't have any other options in MultiFactorProviderConfig.

File: src/auth/auth-config.ts
Comment (prameshj) [inline src/auth/auth-config.ts:694]: Asking for my own clarification - this validation is for the response from the backend and the above code in "validate()" is for user/customer input, correct?

File: src/auth/project-config.ts
Comment (prameshj) [inline src/auth/project-config.ts:65]: Don't you need this field in ProjectConfigClientRequest and ProjectConfigServerResponse?

**Comment by prameshj** [comment]
Also adding @Xiaoshouzi-gh to take a look.

File: src/auth/auth-config.ts
Comment (Xiaoshouzi-gh) [inline src/auth/auth-config.ts:528]: nit: add a space before *

File: test/unit/auth/auth-config.spec.ts
Comment (Xiaoshouzi-gh) [inline test/unit/auth/auth-config.spec.ts:209]: Do you have any tests on these field for the validation you wrote, they could be either located in auth-config.spec or auth.spec

File: src/auth/auth-config.ts
Comment (pragatimodi) [reply src/auth/auth-config.ts:700]: Yes, it is the same thing, just consistent with the rest of the validation in this if statement.

File: test/unit/auth/tenant.spec.ts
Comment (prameshj) [inline test/unit/auth/tenant.spec.ts:124]: nit: multi-factor and provider config are part of the same "multiFactorConfig" struct. So maybe call this:

"'should return the expected server request with multi-factor (SMS, TOTP) and testPhoneNumber config".

I assume thats what phone config means here.

File: test/integration/auth.spec.ts
Comment (prameshj) [inline test/integration/auth.spec.ts:1328]: I think we need one more struct -expectedUpdatedTenantSmsEnabledTotpDisabled to cover the case where TOTP is disabled via provider config, but it preserves the phone MFA settings.

File: test/integration/auth.spec.ts
Comment (prameshj) [inline test/integration/auth.spec.ts:1250]: Shouldn't we add test cases to the project config update flow also , like line 1189?

File: src/auth/auth-config.ts
Comment (pragatimodi) [reply src/auth/auth-config.ts:697]: Yes

File: test/unit/auth/auth-config.spec.ts
Comment (Xiaoshouzi-gh) [inline test/unit/auth/auth-config.spec.ts:376]: Please refactor the buildServerRequest to use one base object, and deepcopy the object in each tests with needed field value updated specifically for the current test.
e.g
```const expectedRequest = { ...};
const expectedDisabledRequest = deepCopy(expectedRequest);
expectedDisabledRequest.state = 'ENABLED';
```
This way it added readability and removed redundant code.

**Review by Xiaoshouzi-gh — CHANGES_REQUESTED** [review]

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts:46]: multiFactorConfig is approved in the API proposal but mfa is the field that corresponds to backend while sending the request on the project side.

File: test/integration/auth.spec.ts
Comment (pragatimodi) [reply test/integration/auth.spec.ts]: This is the integration test suite

File: src/auth/auth-config.ts
Comment (pragatimodi) [reply src/auth/auth-config.ts:632]: Empty array just means no multi factor config which is the default value.

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts:46]: Updated the field to be multiFactorConfig on the user end and translate to mfa for the backend request and response.

File: src/auth/auth-api-request.ts
Comment (pragatimodi) [reply src/auth/auth-api-request.ts:37]: Hiding whitespace changes for now as reverting files with whitespace changes might also revert new code

File: src/auth/auth-config.ts
Comment (lahirumaramba) [inline src/auth/auth-config.ts:517]: Let's add documentation to this and all the new types below:

example see: `MultiFactorConfig`
```
/**
 * Interface representing a multi-factor configuration.
 * This can be used to define whether multi-factor authentication is enabled
 * or disabled and the list of second factor challenges that are supported.
 */
 ```

File: src/auth/auth-config.ts
Comment (lahirumaramba) [inline src/auth/auth-config.ts:518]: I think this might need `/*** ....` style for the documentation to work correctly.

File: src/auth/project-config.ts
Comment (lahirumaramba) [inline src/auth/project-config.ts:81]: It is better if we can keep the strong type here. Can we stick to `ProjectConfigClientRequest` over `any`?

File: src/auth/project-config.ts
Comment (lahirumaramba) [inline src/auth/project-config.ts:125]: I must have missed this, but whey do we have both `multiFactorConfig` and `mfa` in the request type?

**Review by lahirumaramba — CHANGES_REQUESTED** [review]
Thank you! Left a few comments. I will take a look at the tests next. Thanks!

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts:81]: Because of the inconsistent naming of types between backend and SDK, we can't keep a strong type to have flexibility in converting. The same thing has been done in tenant.ts : ```private static validate(request: any, createRequest: boolean): void {```

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts:125]: Because the backend accepts `mfa` but we have `multiFactorConfig` for the field in the SDK to maintain consistency with tenant flow. So this is just a conversion.

File: test/integration/auth.spec.ts
Comment (prameshj) [inline test/integration/auth.spec.ts:1917]: interesting that TOTP gets disabled with a providerConfig that does not have totpProviderConfig{} inside it.

File: test/integration/auth.spec.ts
Comment (pragatimodi) [reply test/integration/auth.spec.ts:1285]: It's not quite clear to me what you mean by this. I think the flow intends to start by updating TOTP and SMS options and then disables TOTP.

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts]: Yes, thanks for catching this miss. Looks like it did not break any tests because the validation function is being tested out separately and this if statement would always be false so the code block wouldn't have been executed. The validKeys has `multiFactorConfig` as the required key in request so the `mfa` field wasn't caught there either.

File: src/auth/project-config.ts
Comment (prameshj) [inline src/auth/project-config.ts:72]: nit: missing "." at the end.

Also, let's add a comment here - "This maps to the "mfa" field in the backend request."

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts:81]: Reopening since this still shows the "any" type. Can we use the UpdateProjectConfigRequest type here?

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts:125]: Reopening since the comment is not here yet. IMO, it is ok to repeat the comment in multiple places.

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts:142]: Super nit: Add a space "/** Backend..." and also no single quote at the end of the comment string.

File: test/integration/auth.spec.ts
Comment (prameshj) [reply test/integration/auth.spec.ts:1290]: Sorry, i must have misread it. I see the test performs this flow as you described.

File: test/integration/auth.spec.ts
Comment (prameshj) [inline test/integration/auth.spec.ts:1917]: nit: can you update the description to "updateTenant() should not disable SMS MFA when TOTP is disabled"

Also, it is slightly inconsistent with the project config tests - there we verify this flow as part of the same test, but we split it here. Just pointing it out, no need to fix it.

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts]: Hmm.. shouldn't it have failed the tests where invalid provider config was being passed in, like the unit tests with `const invalidProviderConfigs = [NaN, 0, 1, '', 'a', {}, { a: 1 }, _.noop, true, false,]`?

**Review by prameshj — COMMENTED** [review]
Mostly looks good to me. Some nits about rewording comments and one question about using a strong type rather than any.

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts:81]: This somehow seems to be getting auto-resolved. @pragatimodi can you look into making this type change?

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts]: @pragatimodi

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts]: This doesn't reflect in the diff yet.

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts:125]: @pragatimodi

File: test/integration/auth.spec.ts
Comment (prameshj) [reply test/integration/auth.spec.ts:1242]: @pragatimodi the diff doesn't show totpProviderConfig still.

File: src/auth/project-config.ts
Comment (prameshj) [reply src/auth/project-config.ts:142]: I think a comment like `'Backend API returns "mfa" in case of project config and "multiFactorConfig" in case of tenant config. The SDK exposes it as multiFactorConfig always.'` would be useful. @pragatimodi

**Review by prameshj — COMMENTED** [review]
Tagged you in a few comments that aren't getting reflected in the diffs yet. LGTM once those are implemented.

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts]: Well, invalidProviderConfigs was not tested because the field itself was marked as `nil` considering the `mfa` values weren't set.

File: src/auth/project-config.ts
Comment (pragatimodi) [reply src/auth/project-config.ts:125]: `link to the docs in the comment` Not sure which doc you are talking about here

Comment (prameshj) [reply src/auth/project-config.ts:127]: i meant https://cloud.google.com/identity-platform/docs/reference/rest/v2/projects.tenants#resource:-tenant and https://cloud.google.com/identity-platform/docs/reference/rest/v2/Config

**Review by prameshj — APPROVED** [review]

File: src/auth/project-config.ts
Comment (Xiaoshouzi-gh) [reply src/auth/project-config.ts:127]: Adding to this thread - the reason why SDK is using `multiFactorConfig ` is because this is the approved name by API council.

**Review by Xiaoshouzi-gh — COMMENTED** [review]
Will LGTM after the  documentation fix.

**Comment by lahirumaramba** [comment]
Thank you, @pragatimodi!
Added `release:stage` to trigger integration tests CI. Please fix the CI issues. Thanks!

File: src/auth/auth-config.ts
Comment (pragatimodi) [reply src/auth/auth-config.ts:1640]: Will address these issues in a different PR. Want to keep this one related to MFA changes.

File: src/auth/auth-config.ts
Comment (pragatimodi) [reply src/auth/auth-config.ts:705]: It would not be null. I think adjIntervals will not be defined in that case which is being checked in the next line. Thanks for double checking!

File: etc/firebase-admin.auth.api.md
Comment (pragatimodi) [reply etc/firebase-admin.auth.api.md:268]: Good point. Yes these types don't exist in the API proposal. Should we update it? @prameshj @Xiaoshouzi-gh @renkelvin

Comment (renkelvin) [reply etc/firebase-admin.auth.api.md:268]: Yes, we should stick to the API proposals.

Maybe just make MultiFactorAuthConfig internal?

File: src/auth/project-config.ts
Comment (Xiaoshouzi-gh) [inline src/auth/project-config.ts:128]: nit: space after //

File: etc/firebase-admin.auth.api.md
Comment (Xiaoshouzi-gh) [reply etc/firebase-admin.auth.api.md:268]: to make it internal, please check `SmsRegionsAuthConfig` for reference.

Comment (pragatimodi) [reply etc/firebase-admin.auth.api.md:268]: This class isn't user facing. It just implements the MultiFactorConfig interface which is public. @lahirumaramba should i remove /** */ documentation to undo this?

Comment (lahirumaramba) [reply etc/firebase-admin.auth.api.md]: Since we export it, it will still show up in the public API. I think we can mark it as `@internal` similar to here: https://github.com/firebase/firebase-admin-node/blob/81672d6e67bf36008fd9ecd4921e7ed4269b0ff6/src/auth/auth-config.ts#L1520

Comment (pragatimodi) [reply etc/firebase-admin.auth.api.md]: Updated.

**Comment by Xiaoshouzi-gh** [comment]
LGTM, waiting on the test cases explanation docs (internal).

File: src/auth/auth-config.ts
Comment (prameshj) [reply src/auth/auth-config.ts:1644]: Looks like this has been addressed.. Can this be resolved, @pragatimodi ?

File: src/auth/project-config.ts
Comment (prameshj) [inline src/auth/project-config.ts:165]: "deepCopy(this.multiFactorConfig_)" ?

Comment (pragatimodi) [reply src/auth/project-config.ts:165]: Since `multiFactorConfig` is the getter function, shouldn't we use it that way to fetch `multiFactorConfig_` values in that?

**Review by Xiaoshouzi-gh — APPROVED** [review]

**Review by lahirumaramba — APPROVED** [review]
LGTM! Thank you!

**Please get the docs reviewed before you merge this.** @kevinthecheung might be able to help. Thanks!

File: src/auth/auth-config.ts
Comment (kevinthecheung) [inline src/auth/auth-config.ts:529]:
```suggestion
   * TOTP multi-factor provider config.
```

File: src/auth/auth-config.ts
Comment (kevinthecheung) [inline src/auth/auth-config.ts:540]:
```suggestion
    *  to compensate for clock skew.
```

File: src/auth/project-config.ts
Comment (kevinthecheung) [inline src/auth/project-config.ts:73]:
```suggestion
  /**
   * The project's multi-factor auth configuration.
   * Supports only phone and TOTP.
   */
```

**Review by kevinthecheung — CHANGES_REQUESTED** [review]

File: src/auth/auth-config.ts
Comment (pragatimodi) [reply src/auth/auth-config.ts:507]: This seems to be correct. `factorIds` only supports phone right now.

**Review by kevinthecheung — APPROVED** [review]

**Review by renkelvin — APPROVED** [review]

**Pull request merged**
